spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/pubsub/GcpPubSubEmulatorConfiguration.java

```diff
@@ -0,0 +1,55 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.pubsub;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.api.gax.core.NoCredentialsProvider;
+import com.google.api.gax.grpc.GrpcTransportChannel;
+import com.google.api.gax.rpc.FixedTransportChannelProvider;
+import com.google.api.gax.rpc.TransportChannelProvider;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.cloud.gcp.autoconfigure.core.GcpContextAutoConfiguration;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * @author Andreas Berger
+ */
+@Configuration
+@ConditionalOnProperty("PUBSUB_EMULATOR_HOST")
+@AutoConfigureBefore(GcpContextAutoConfiguration.class)
+public class GcpPubSubEmulatorConfiguration {
```

Comment:
What is the `Emulator` in the `main` code?

I would suggest to move this to something like `testing-support` if there is indeed a requirement to make this feature public.

I would stay away in production ready API of this kind of tools.

Also an absence of JavaDocs makes me nervous as well.

Reply:
The emulator is not only used for testing but also for local setups. A switch that applies when an emulator environment is detected is exactly what spring auto-configuration is supposed to do, isn't it?

Reply:
Yeah... I see now what you mean.

That sounds like more like `Embedded`, but I agree that `emulator` is better name because a service is external.

I think we need special `@ConfigurationProperties` on the matter instead of that ugly ENV var (or not?).

Shouldn't we also have a `@Conditional` to be sure that emulator is definitely run on that host?

Why is this `Before(GcpContextAutoConfiguration.class)` ?

And JavaDocs, please, for better understanding

Reply:
Local runs are used in a testing context, right?
I'd be hard pressed to think anyone is going to use the emulator to run production apps.
So it could still make sense to add this to some kind of `testing-support` package. Thoughts?

Reply:
@artembilan I called it Emulator b/c it is also called like this by Google. And it is run as a local console app. What do you mean with "special `@ConfigurationProperties`"? The name is [predefined by Google es well](https://cloud.google.com/pubsub/docs/emulator?hl=en#manually_setting_the_variables). The existence of this Property should be enough to check for. The `@Before` annotation was added to ensure that the Credentials Provided by this Configuration has precedence over the one defined by `GcpContextAutoConfiguration`.

@joaoandremartins As I have already said it is not only testing, it is also used for local dev setup, so a separat dependency would not make sense.

Reply:
I mean Spring Boot's [Configuration by Convention](https://docs.spring.io/spring-boot/docs/2.0.0.RC1/reference/htmlsingle/#boot-features-external-config). 
That way we have a lot of possibilities to configure the same value.

Reply:
So you want to ignore the existing name and create an own spring property for this? 

Currently, the entire configuration of the pub/sub client is provided externally. No extra configuration is required within the Google Cloud, as all relevant information are provided through environment variables. For the local dev setup the call `gcloud beta emulators pubsub env-init` would be sufficient to activate the emulator.

Reply:
That's not how Spring Boot auto-configuration works.
You can't assume that that ENV var is always going to be there and on all the platforms.

That's why Spring Boot introduces conventional properties - to let to end-user to provide it either way.

You always can remap ENV var to the conventional properties.

That's sad that I don't have enough arguments to convince you, but I definitely would go the best practice from exiting Spring Boot experience.

Reply:
+1 on having properties to configure the emulator instead of envvars.
From what I could see, it's definitely the Spring Boot best practice.